scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -1516,6 +1536,152 @@
+                         :resources pool-specific-resources}})))))
+
+(defn handle-kubernetes-scheduler-pool
+  "Handle scheduling pending jobs using the Kubernetes Scheduler."
+  [conn pool-name->pending-jobs-atom pool-name scheduler-config compute-clusters
+   job->acceptable-compute-clusters-fn user->quota user->usage-future mesos-run-as-user]
+  (log-structured/info "Running handler for Kubernetes Scheduler pool" {:pool pool-name})
+  (try
+    (let [user->usage (tracing/with-span [s {:name "scheduler.kubernetes-handler.resolve-user-to-usage-future"
+                                             :tags {:pool pool-name :component tracing-component-tag}}]
+                        @user->usage-future)
+          {max-considerable :max-jobs-considered} scheduler-config
+          db (db conn)
+          pending-jobs (get @pool-name->pending-jobs-atom pool-name)
+          ;; We need to filter pending jobs based on quota so that we don't
+          ;; submit beyond what users have quota to actually run.
+          jobs (prom/with-duration
+                 prom/scheduler-handle-resource-offers-pending-to-considerable-duration {:pool pool-name}
+                 (timers/time!
+                  (timers/timer (metric-title "kubernetes-handler-considerable-jobs-duration" pool-name))
+                  (pending-jobs->considerable-jobs
+                   db pending-jobs user->quota user->usage max-considerable pool-name)))
+          job-uuids (set (map :job/uuid jobs))]
+      (log-structured/info "Considering jobs" {:pool pool-name :number-considered-jobs (count job-uuids)})
+      (if (seq jobs)
+        (do
+          (swap! pool-name->pending-jobs-atom
+                 remove-matched-jobs-from-pending-jobs
+                 job-uuids pool-name)
+          (log-structured/debug (print-str "Updated pool-name->pending-jobs-atom:" (get @pool-name->pending-jobs-atom pool-name))
+                                {:pool pool-name})
+          (let [autoscaling-compute-clusters (filter #(cc/autoscaling? % pool-name) compute-clusters)
+                ;; Distribute each job to a compute cluster
+                ;; TODO(alexh): confirm this is random
+                compute-cluster->jobs (distribute-jobs-to-compute-clusters
+                                       jobs pool-name autoscaling-compute-clusters
+                                       job->acceptable-compute-clusters-fn)]
+            (log-structured/info "Starting launch directly to Kubernetes" {:pool pool-name})
+            (->> compute-cluster->jobs
+                 (map
+                  (fn [[compute-cluster jobs]]
+                    (let [kill-lock-object (cc/kill-lock-object compute-cluster)
+                          task-metadata-seq (jobs->kubernetes-task-metadata jobs pool-name mesos-run-as-user compute-cluster)
+                          task-txns (map (fn [job metadata] (job->task-txn job metadata compute-cluster)) jobs task-metadata-seq)]
+                      (try
```

Comment:
The overall shape of this is very close to `launch-tasks-for-cluster` with the token spending code inlined in from `launch-matches!`.

Can you refactor the common components out and call from here and existing place?

Reply:
This is the one refactor I totally agree needs to happen and would be good, but is a bit complex. I will add it to the "master list of TODOs" I'm developing for you tomorrow.